File: Q4.py
import numpy as np
import matplotlib.pyplot as plt
f=open("Q_4.txt","r")           #Read the data obtained from C.
Data=f.readlines()
i=0
N=64                            #This N should be the same as used in the corresponding C code.
k=np.zeros(N) 
fk=np.zeros(N)
for D in Data:
    D1,D2=D.split()
    k[i]=float(D1)
    fk[i]=float(D2)
    i=i+1

x=np.linspace(-5,5,N)           #The x_min and x_max should be the same as in the corresponding C code.
fx=np.exp(-x**2)                #The function whose Fourier transform is required.
plt.subplot(1,2,1)
plt.plot(x,fx,'b')
plt.xlabel(r'$x$',fontsize=16)
plt.ylabel(r'$f(x)$',fontsize=16)

plt.grid(True)

plt.subplot(1,2,2)
plt.plot(k,fk,'b')                              #From the C code using FFT.
plt.plot(k,np.sqrt(0.5)*np.exp(-k**2/4),'r')    #Analytical result.
plt.xlabel(r'$k$',fontsize=16)
plt.ylabel(r'$\tilde{f}(k)$',fontsize=16)
plt.legend(['From C (using FFT)','Analytical'],fontsize=16)

plt.grid(True)
plt.show()

File: Q4.c
//Question 4 (First part): Fourier transform of Gaussian using FFT algorithm.
//The plotting part is in a python file Q4.py
#include <stdio.h>    
#include <complex.h>
#include <math.h>

double _Complex DFT(double F[],int n, float Q)	//FFT algorithm
{
	if(n!=1)
	{
		double ev[n/2];
		double od[n/2];
		for(int j=0;j<n/2;j++)
		{
			ev[j]=F[2*j];
		}
		for(int j=0;j<n/2;j++)
		{
			od[j]=F[2*j+1];
		}
		return DFT(ev,n/2,Q)+cexp(-2*M_PI*I*Q/n)* DFT(od,n/2,Q);
	}
	else
	{
		return F[0];
	}
}

int main() 
{
	int N=64;
	double f[N];
	float x_min=-5;
	float x_max=5;
	float Dx=(x_max-x_min)/(N-1);
	for(int i=0;i<N;i++)
	{
		f[i]=exp(-pow(x_min+i*Dx,2));	//Function whose Fourier transform is required.
	}
	
	double _Complex Fourier[N];
	double k[N];
	float q;
	FILE *data;
	data = fopen("Q_4.txt","w");
	if (!data)
	perror("fopen");
	
	for(int i=0;i<N;i++)	//Print data to a text file.
	{
		q=-N/2+i;			//Following the convention used by python for k values.
		k[i]=2*M_PI*q/(N*Dx);
		Fourier[i]=Dx*sqrt(1/(2*M_PI))*cexp(-I*k[i]*x_min)*DFT(f,N,q);
		fprintf(data,"%f   %f\n",k[i],creal(Fourier[i]));
	}
	fclose(data);
}
//To compile: gcc -Wall Q4.c -lm -lgsl -lgslcblas -o Q4
//To run	: ./Q4
